File: YeChanSong_코딩테스트_연습/DFS_BFS/여행경로.py
import sys
sys.setrecursionlimit(100000)

def buildGraph():
    global data, adj
    
    for s,d in data:
        if adj.get(s) == None:
            adj.setdefault(s,[d])
        else:
            adj[s].append(d)
        if adj.get(d) == None:
            adj.setdefault(d,[])
        
def dfs(name):
    global adj, hist
    
    adj[name] = sorted(adj[name],reverse=True)
    while len(adj[name]):
        nxt = adj[name].pop()
        dfs(nxt)
    hist.append(name)    
    

def solution(tickets):
    answer = []
    global data, adj, hist
    adj = dict()
    hist = list()
    data = tickets
    buildGraph()
    dfs("ICN")
    hist.reverse()
    answer = hist
    
    return answer

File: YeChanSong_코딩테스트_연습/DFS_BFS/여행경로.java
import java.util.*;

class Solution {
    static HashMap<String, LinkedList<String>> adj = new HashMap<>();
    static ArrayList<String> hist = new ArrayList<>();
    public String[] solution(String[][] init) {
        String[] answer;
        Arrays.sort(init, (String[] a, String[] b)->{
            if(a[0].equals(b[0]))
                return a[1].compareTo(b[1]);
            else
                return a[0].compareTo(b[0]);
        });
        for(String[] str: init){
            if(adj.get(str[0]) == null)
                adj.put(str[0], new LinkedList<>());
            if(adj.get(str[1]) == null)
                adj.put(str[1], new LinkedList<>());
            adj.get(str[0]).add(str[1]);
        }
        dfs("ICN");
        answer = new String[hist.size()];
        for(int i=0;i<hist.size();i++)
            answer[i] = hist.get(hist.size()-1-i);
        return answer;
    }
    public void dfs(String start){
        
        String tmp;
        while(!adj.get(start).isEmpty()){
            tmp = adj.get(start).poll();
            dfs(tmp);
        }
        hist.add(start);
    }
}